Fix collection_singular_ids= bug

composerinteralia opened the pull request:
When the association's primary key is set as a symbol,
`CollectionAssociation#ids_writer` fails to cast the ids to integers.
This is because `AssociationReflection#association_primary_key_type`
and `ThroughReflection#association_primary_key_type` rely on
`ModelSchema.type_for_attribute`, which only accepts a string.
The result is an `ActiveRecord::RecordNotFound` error.

Fixes #27864

composerinteralia committed:
Fix collection_singular_ids= bug

When the association's primary key is manually set as a symbol and called with an array of strings, CollectionAssociation#ids_writer
fails to cast the ids to integers. This is because
AssociationReflection#association_primary_key_type and
ThroughReflection#association_primary_key_type return the incorrect
type, since ModelSchema.type_for_attribute only accepts a string. The
result is an ActiveRecord::RecordNotFound error.

rafaelfranca commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+*   Fix `association_primary_key_type` for reflections with symbol primary key
+
+    Fixes #27864
+
```
Do you mind to add your name here?

composerinteralia replied:
sorry about that!

composerinteralia replied:
Done!

rafaelfranca reviewed (changes requested):
Thank you so much for the PR. I comment in the CHANGELOG to give you proper recognition of your change.

composerinteralia commented:
Oops, I can see now that my tests are no good. I will take another look. @rafaelfranca Is it better to keep this open or to close for now and open a new PR once I have this figured out?

rafaelfranca commented:
Let's keep this open

kamipo commented on activerecord/test/cases/reflection_test.rb:
```diff
@@ -335,6 +335,15 @@ def test_association_primary_key
+    assert_equal ActiveModel::Type::String,  Author.reflect_on_association(:essay).association_primary_key_type.class
+
+    # Through Association
+    assert_equal ActiveModel::Type::String, Author.reflect_on_association(:essay_category).association_primary_key_type.class
```
How about `assert_equal :string, Author.reflect_on_association(:essay_category).association_primary_key_type.type`?

composerinteralia replied:
Yeah, that makes sense to me. Thanks for your help!

composerinteralia commented:
@rafaelfranca I fixed my errors and rebased on master, but now I am getting new errors that seem to be related to this: https://github.com/rails/rails/commit/aa647b46cce55ec12f5895e403c0d1b85502c8e0
Any advice?

composerinteralia commented:
I can't recreate the failure on my end. Anyone want to help?

eileencodes commented:
I couldn't reproduce locally so I restarted the build and this time it passed. I think it was unrelated to your changes.

eileencodes commented:
Backported to 5-0-stable in https://github.com/rails/rails/commit/93a44497ed7b28f5691a5cb1685261af5dcf64c1
